lib/variable/reduction.cpp

```diff
@@ -22,148 +23,130 @@ using namespace scipp::core;
-  return reduce_idempotent(var, dim, core::element::logical_and_equals,
-                           FillValue::True, "all");
+  return reduce_dim(var, dim, all_into, FillValue::True);
 }
 
 /// Return the maximum along given dimension.
 ///
 /// Variances are not considered when determining the maximum. If present, the
 /// variance of the maximum element is returned.
 Variable max(const Variable &var, const Dim dim) {
-  return reduce_idempotent(var, dim, core::element::max_equals,
-                           FillValue::Lowest, "max");
+  return reduce_dim(var, dim, max_into, FillValue::Lowest);
 }
 
 /// Return the maximum along given dimension ignoring NaN values.
 ///
 /// Variances are not considered when determining the maximum. If present, the
 /// variance of the maximum element is returned.
 Variable nanmax(const Variable &var, const Dim dim) {
-  return reduce_idempotent(var, dim, core::element::nanmax_equals,
-                           FillValue::Lowest, "nanmax");
+  return reduce_dim(var, dim, nanmax_into, FillValue::Lowest);
 }
 
 /// Return the minimum along given dimension.
 ///
 /// Variances are not considered when determining the minimum. If present, the
 /// variance of the minimum element is returned.
 Variable min(const Variable &var, const Dim dim) {
-  return reduce_idempotent(var, dim, core::element::min_equals, FillValue::Max,
-                           "min");
+  return reduce_dim(var, dim, min_into, FillValue::Max);
 }
 
 /// Return the minimum along given dimension ignorning NaN values.
 ///
 /// Variances are not considered when determining the minimum. If present, the
 /// variance of the minimum element is returned.
 Variable nanmin(const Variable &var, const Dim dim) {
-  return reduce_idempotent(var, dim, core::element::nanmin_equals,
-                           FillValue::Max, "nanmin");
+  return reduce_dim(var, dim, nanmin_into, FillValue::Max);
+}
+
+Variable mean_impl(const Variable &var, const Dim dim, const Variable &count) {
+  return normalize_impl(sum(var, dim), count);
+}
+
+Variable nanmean_impl(const Variable &var, const Dim dim,
+                      const Variable &count) {
+  return normalize_impl(nansum(var, dim), count);
+}
+
+namespace {
+template <class... Dim> Variable count(const Variable &var, Dim &&... dim) {
+  if (!is_bins(var)) {
+    if constexpr (sizeof...(dim) == 0)
+      return var.dims().volume() * units::none;
+    else
+      return ((var.dims()[dim] * units::none) * ...);
+  }
+  if (const auto mask = variableFactory().irreducible_event_mask(var);
+      mask.is_valid()) {
+    // Trick to get the sizes of bins if masks are present - bin the masks
+    // using the same dimension & indices as the data, and then sum the
+    // inverse of the mask to get the number of unmasked entries.
+    return sum(make_bins_no_validate(var.bin_indices(),
+                                     variableFactory().elem_dim(var), ~mask));
+  }
```

Comment:
This branch ignores `dim`. I think this was fine while event masks where supported only in, e.g., `da.bins.mean()`, but now `da.mean(dim)` will also handle them... and this is miscounting now. I think the computation of the mean later on would actually throw though, since the returned sum has extra dims?